packages/core/src/workflow/workflow.ts

```diff
@@ -0,0 +1,221 @@
+import { Context } from "./context";
+import {
+  type EventTypes,
+  StartEvent,
+  StopEvent,
+  WorkflowEvent,
+} from "./events";
+
+export type StepFunction<T extends WorkflowEvent = WorkflowEvent> = (
+  context: Context,
+  ev: T,
+) => Promise<WorkflowEvent | void>;
+
+type EventTypeParam = EventTypes | EventTypes[];
+
+export class Workflow {
+  #steps: Map<
+    StepFunction<any>,
+    { inputs: EventTypes[]; outputs: EventTypes[] | undefined }
+  > = new Map();
+  #contexts: Set<Context> = new Set();
+  #verbose: boolean = false;
+  #timeout: number | null = null;
+  #validate: boolean = false;
+
+  constructor(
+    params: {
+      verbose?: boolean;
+      timeout?: number;
+      validate?: boolean;
+    } = {},
+  ) {
+    this.#verbose = params.verbose ?? false;
+    this.#timeout = params.timeout ?? null;
+    this.#validate = params.validate ?? false;
+  }
+
+  addStep<T extends WorkflowEvent>(
```

Comment:
For some reason, my brain really wants `addStep` to have the step as the first argument 😅